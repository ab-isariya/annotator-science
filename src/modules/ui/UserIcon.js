import {forwardRef} from 'react';
import PropTypes from 'prop-types';

import {Link} from 'react-router-dom';
import {useOktaAuth} from '@okta/okta-react';

import Stack from '@modules/ui/Stack';
import {useUser} from '@user';

import {ReactComponent as Dashboard} from '@assets/svgs/Dashboard.svg';
import {ReactComponent as Help} from '@assets/svgs/Help.svg';
import {ReactComponent as Logout} from '@assets/svgs/Logout.svg';
import {ReactComponent as User} from '@assets/svgs/User.svg';

const UserIcon = forwardRef(({canShowDropdown, onClick, children}, ref) => {
  const {data: user} = useUser();
  const {oktaAuth} = useOktaAuth();

  const dropdown = (
    // z-50 class is not working
    <div
      className="absolute left-0 w-60 rounded-md shadow-sm bg-white"
      style={{zIndex: 100}}>
      <div
        className=""
        role="menu"
        aria-orientation="vertical"
        aria-labelledby="options-menu">
        <Link to="/" role="menuitem">
          <div className="flex items-center space-x-2.5 p-2.5 hover:bg-grey-50">
            <Dashboard /> <div>ScienceIO Dashboard</div>
          </div>
        </Link>

        <Link to="/" role="menuitem">
          <div className="flex items-center space-x-2.5 p-2.5 hover:bg-grey-50">
            <Help /> <div>Help</div>
          </div>
        </Link>

        <div
          className="flex items-center space-x-2.5 p-2.5 hover:bg-grey-50"
          onClick={() => {
            oktaAuth.signOut('/');
            location.reload();
          }}>
          <Logout />
          <div>Logout</div>
        </div>
      </div>
    </div>
  );

  const email = user?.primary_email;

  if (!email)
  {
    return (
      <></>
    );
  }

  return (
    <div className="flex space-x-2.5 items-center">
      {/* TODO BUG: without style, this div is taller than icon */}
      <div className="relative" ref={ref} style={{height: '45px'}}>
        <Stack className={onClick ? 'cursor-pointer' : ''} onClick={onClick}>
          <User />
          <span className="text-base capitalize">
            {email.charAt(0) || ''}
          </span>
        </Stack>
        {canShowDropdown && dropdown}
      </div>

      <div className="text-base capitalize">
        <span className="font-light">{email.split('@')[0]}</span>
        <br />
        {children}
      </div>
    </div>
  );
});

UserIcon.displayName = 'UserIcon';

UserIcon.propTypes = {
  /** Dropdown component to display when clicking on icon */
  canShowDropdown: PropTypes.bool,
  /** Function to trigger display of dropdown component */
  onClick: PropTypes.func,
  //React node children
  children: PropTypes.node
};

export default UserIcon;
