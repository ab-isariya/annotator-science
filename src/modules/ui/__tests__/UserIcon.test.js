import UserIcon from '../UserIcon';
import renderer from 'react-test-renderer';

jest.mock('@okta/okta-react', () => ({
  useOktaAuth: () => {
    return {
      authState: {isAuthenticated: true},
      authService: {handleAuthentication: jest.fn()},
      oktaAuth: {
        tokenManager: {
          on: jest.fn()
        }
      }
    };
  }
}));

jest.mock('@user', () => ({
  useUser: () => {
    return {
      data: {
        primary_email: 'hello@test.com'
      }
    };
  }
}));

describe('<UserIcon/>', () => {
  it('Renders without crashing', () => {
    const comp = renderer.create(<UserIcon />).toJSON();

    expect(comp).toMatchSnapshot();
  });
});
