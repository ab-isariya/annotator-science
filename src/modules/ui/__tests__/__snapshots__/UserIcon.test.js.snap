// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`<UserIcon/> Renders without crashing 1`] = `
<div
  className="flex space-x-2.5 items-center"
>
  <div
    className="relative"
    style={
      Object {
        "height": "45px",
      }
    }
  >
    <div
      className=" relative"
    >
      <svg>
        User.svg
      </svg>
      <div
        className="absolute top-2/4 left-2/4"
        style={
          Object {
            "transform": "translate(-50%, -50%)",
          }
        }
      >
        <span
          className="text-base capitalize"
        >
          h
        </span>
      </div>
    </div>
  </div>
  <div
    className="text-base capitalize"
  >
    <span
      className="font-light"
    >
      hello
    </span>
    <br />
  </div>
</div>
`;
